File: src/core/validators/Interchange/InterchangeMustNotDuplicateIdentityTemplateName.cs
using System.Linq;
using MetaEd.Grammar.Antlr;

namespace MetaEd.Core.Validator.Interchange
{
    public class InterchangeMustNotDuplicateIdentityTemplateName : ValidationRuleBase<MetaEdGrammar.InterchangeContext>
    {
        private readonly ISymbolTable _symbolTable;

        public InterchangeMustNotDuplicateIdentityTemplateName(ISymbolTable symbolTable)
        {
            _symbolTable = symbolTable;
        }

        private static string[] DuplicateIdentityTemplates(MetaEdGrammar.InterchangeContext context)
        {
            var identityTemplates = context.interchangeComponent().interchangeIdentityTemplate().Select(x => x.ID().GetText());

            //group and filter duplicates
            return identityTemplates.GroupBy(x => x).Where(group => group.Count() > 1).Select(group => group.Key).ToArray();
        }

        public override bool IsValid(MetaEdGrammar.InterchangeContext context)
        {
            return !DuplicateIdentityTemplates(context).Any();
        }

        public override string GetFailureMessage(MetaEdGrammar.InterchangeContext context)
        {
            var identifier = context.interchangeName().GetText();
            var duplicateIdentityTemplates = DuplicateIdentityTemplates(context);

            return string.Format("Interchange '{0}' declares duplicate identity template{2} '{1}'.", identifier, string.Join("', '", duplicateIdentityTemplates), duplicateIdentityTemplates.Count() > 1 ? "s" : string.Empty);
        }
    }
}

File: src/core/validators/Interchange/InterchangeMustNotDuplicateIdentityTemplateName.ts
import { ValidationRuleBase } from "../ValidationRuleBase";
import {ISymbolTable} from '../SymbolTable'
export class InterchangeMustNotDuplicateIdentityTemplateName extends ValidationRuleBase<MetaEdGrammar.InterchangeContext>
{
    private symbolTable: ISymbolTable;
    constructor(symbolTable: ISymbolTable) {
        super();
        this.symbolTable = symbolTable;
    }
    private static duplicateIdentityTemplates(context: MetaEdGrammar.InterchangeContext): string[] {
        let identityTemplates = context.interchangeComponent().interchangeIdentityTemplate().map(x => x.ID().getText());
        return identityTemplates.GroupBy(x => x).filter(group => group.Count() > 1).map(group => group.Key).ToArray();
    }
    public isValid(context: MetaEdGrammar.InterchangeContext): boolean {
        return InterchangeMustNotDuplicateIdentityTemplateName.duplicateIdentityTemplates(context).length == 0;
    }
    public getFailureMessage(context: MetaEdGrammar.InterchangeContext): string {
        let identifier = context.interchangeName().getText();
        let duplicateIdentityTemplates = InterchangeMustNotDuplicateIdentityTemplateName.duplicateIdentityTemplates(context);
        return `Interchange '${identifier}' declares duplicate identity template${duplicateIdentityTemplates.length > 1 ? "s" : ""} '${duplicateIdentityTemplates.join(', ')}'.`; 
    }
}
